### PR TITLE
fix: Allow serving CRs and CRDs and other improvements

### PR DESCRIPTION
**Pull request by banjoh**
Here are the list of changes made to achieve being able to read custom resources. I also added some improvements in the process

* Use [`unstructured`](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured#Unstructured) objects whenever possible especially when concrete types are unknown.
* Add middleware to dump requests and responses. These dumps are logged whenever you use `--debug` CLI option. They are handy when developing.
* Add getVersion API to allow running `kubectl cluster-info`. Some tools like `k9s` use this to get server info
* Always set `selfsubjectaccessreviews` to `status.allowed=true` so as to authorise all requests. `k9s` for example needs this
* For all listing API calls, respond with empty lists not 404s. This makes `kubectl` behave as expected whenever there are no resources found
```sh
[evans] $ kubectl get pods -n not-there
No resources found in not-there namespace.
``` 
* Better handling of Table responses by respecting groups `Accept` header values

Fixes: #48 
Partially addresses #75

### CODE REVIEW

**Comment by banjoh** [comment]
As a side effect of these changes, I got `k9s` working. I have not testing it much but some basics work already.

Screenshot 2023-10-12 at 15 32 10

File: pkg/api/server.go [hunk 1]
```diff
@@ -60,21 +63,28 @@ var (
 	}
 )
 
-type handler struct {
-	clusterData sbctl.ClusterData
-}
+type (
```
Comment (divolgin) [inline pkg/api/server.go:66]: consistency

```
type <name> struct
```

File: pkg/api/server.go [hunk 2]
```diff
@@ -266,38 +302,51 @@ func (h handler) getAPIV1ClusterResources(w http.ResponseWriter, r *http.Request
...
+		if !fileExists(fileName) {
+			obj := unstructured.UnstructuredList{}
+			obj.SetGroupVersionKind(schema.GroupVersionKind{
+				Group:   mux.Vars(r)["group"],
+				Version: mux.Vars(r)["version"],
+				Kind:    resource,
+			})
+
+			result = &obj
+			break
```
Comment (divolgin) [inline pkg/api/server.go:322]: This should do nothing and continue instead.

File: pkg/api/server.go [hunk 3]
```diff
@@ -806,23 +875,57 @@ func (h handler) getAPIsClusterResources(w http.ResponseWriter, r *http.Request)
...
```
Comment (divolgin) [inline pkg/api/server.go:916]: Do nothing and continue. Set result to empty object if no data was found in the entire loop.

File: pkg/api/middleware.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+func (w *requestResponseDumper) Flush() {
+	w.ResponseWriter.(http.Flusher).Flush()
+}
+
+func (w *requestResponseDumper) Hijack() (net.Conn, *bufio.ReadWriter, error) {
+	return w.ResponseWriter.(http.Hijacker).Hijack()
+}
+
+func logObject(prefix string, o any) {
```
[the remark this answers is not in the document]
Comment (banjoh) [reply pkg/api/middleware.go:68]: I've found `any` a clearer type to use but I'll revert in favour of consistency

File: pkg/sbctl/compatibility.go [hunk 5]
```diff
@@ -22,21 +24,47 @@ func init() {
...
 	}
 
-	decoded, gvk, err = decode(data, nil, nil)
-	if err != nil {
-		return nil, nil, errors.Wrap(err, "failed to decode wrapped data")
+	if decoded == nil {
+		// Try to decode object into an unstructured object
+		var v unstructured.Unstructured
+		err = json.Unmarshal(originalData, &v)
+		if err == nil {
```
Comment (divolgin) [inline pkg/sbctl/compatibility.go:49]: value of `err` changes a few times in this block.  how about  logging it every time

```
if err != nil {
   // log and continue
}
```

File: pkg/api/utils.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
+package api
+
+import (
+	"strings"
+)
+
+// parseAcceptHeader parses the a header and returns a map of key value pairs
+// An kubectl Accept header for example looks like this:
+// application/json;as=Table;g=meta.k8s.io;v=v1beta1, application/json
+func parseAcceptHeader(headers []string) map[string]string {
```
[the remark this answers is not in the document]
Comment (banjoh) [reply pkg/api/utils.go:10]: I'll search around for any existing implementation

**Review by divolgin — APPROVED** [review]

**Pull request merged**